rust/lance-encoding/src/encodings/physical/bitpack.rs

```diff
@@ -0,0 +1,220 @@
+                
+                // current source byte being written to destination
+                let mut curr_src = src[src_idx] >> src_offset & curr_mask as u8;
+                
+                // how many bits from the current source value have been written to destination
+                let mut src_bits_written = 0;
+
+                // the offset within the current destination byte to write to
+                let mut dst_offset = 0;
+
+                while src_bits_written < self.bits_per_value {
+                    // write bits from current source byte into destination
+                    dst[dst_idx] += (curr_src) << dst_offset;
+                    let bits_written = (self.bits_per_value - src_bits_written)
+                        .min(8 - src_offset)
+                        .min(8 - dst_offset);
+                    src_bits_written += bits_written;
+                    dst_offset += bits_written;
+                    src_offset += bits_written;
+
+                    if dst_offset == 8 {
+                        dst_idx += 1;
+                        dst_offset = 0;
+                    }
+
+                    if src_offset == 8 {
+                        src_idx += 1;
+                        src_offset = 0;
+                        curr_mask >>= 8;
+                        if src_idx == src.len() {
+                            break;
+                        }
+                        curr_src = src[src_idx] & curr_mask as u8;
+                    }
+                }
+
+                // advance destination offset to the next location
+                // note that we don't need to do this if we wrote the full number of bits
+                // because source index would have been advanced by the inner loop above
+                if self.uncompressed_bits_per_value != self.bits_per_value {
+                    let mut partial_bytes_written = self.bits_per_value / 8;
+
+                    // if we didn't write the full byte for the last byte, increment by one because
+                    // we wrote a partial byte
+                    if self.uncompressed_bits_per_value % self.bits_per_value != 0 {
+                        partial_bytes_written += 1;
+                    }
+                    dst_idx += (byte_len as u64 - partial_bytes_written + 1) as usize;
+                }
+            }
+        }
+    }
+
+    fn num_buffers(&self) -> u32 {
+        // TODO ask weston what this is about
+        1
+    }
+}
+
+#[cfg(test)]
+pub mod test {
+    use super::*;
+
+    #[test]
+    fn test_decode_into() {
+        let src_buffer = vec![0b11_010_001, 0b0_101_100_0, 0b111_11];
+        let unit = BitpackedPageDecoder {
+            bits_per_value: 3,
+            uncompressed_bits_per_value: 32,
+            data: vec![Bytes::copy_from_slice(&src_buffer)],
+        };
+
+        let mut dest = vec![BytesMut::new()];
+        unit.decode_into(0, 7, &mut dest);
+
+        println!("{:?}", dest);
```

Comment:
Nit: convert to an assert when ready to move out of draft.

Reply:
I deleted this.. we have other tests covering this code path